File: parse_amazon_xml.py
# -*- coding:utf-8 -*-
import time
from lxml import objectify


class ImageInfo:
    def __init__(self):
        self.url = ''
        self.width = ''
        self.height = ''

class BookInfo:
    def __init__(self):
        self.asin = ''
        self.title = ''
        self.binding = ''
        self.author = ''
        self.publisher = ''
        self.publicationDate = ''
        self.images = {}


def getText(dom, tag):
    return getattr(dom, tag).text if tag in dom else ''


def parseXmls(xmls):
    bookInfos = []
    for xml in xmls:
        dom = objectify.fromstring(xml)
        for item in dom.Items.Item:
            bookInfo = BookInfo()
            bookInfo.asin = item.ASIN.text

            attr = item.ItemAttributes
            bookInfo.title = getText(attr, 'Title')
            bookInfo.binding = getText(attr, 'Binding')
            bookInfo.author = getText(attr, 'Author')
            bookInfo.publisher = getText(attr, 'Publisher')
            bookInfo.publicationDate = getText(attr, 'PublicationDate')

            imageLabels = ['SmallImage', 'MediumImage', 'LargeImage']
            for imageLabel in imageLabels:
                image = ImageInfo()
                if imageLabel in item:
                    image.url = getattr(item, imageLabel).URL.text
                    image.width = int(getattr(item, imageLabel).Width.text)
                    image.height = int(getattr(item, imageLabel).Height.text)
                bookInfo.images[imageLabel] = image

            bookInfos.append(bookInfo)

    return bookInfos


def getXmls():
    xmls = []
    for i in range(0, 1440+1, 10):
        path = 'xmls/{}.xml'.format(i)
        with open(path, 'r') as f:
            xml = f.read()
            xmls.append(xml)
    return xmls


def main():
    xmls = getXmls()
    start = time.time()
    bookInfos = parseXmls(xmls)
    end = time.time()
    print('xml数: {}'.format(len(xmls)))
    print('book数: {}'.format(len(bookInfos)))
    print('parse時間: {}秒'.format(end - start))


if __name__ == '__main__':
    main()

File: parse_amazon_xml.rb
#! /usr/bin/env ruby
# coding: utf-8
# frozen_string_literal: true

require 'rexml/document'

class ImageInfo
  attr_accessor :url, :width, :height
end

class BookInfo
  attr_accessor :asin, :title, :binding, :author, :publisher, :publicationDate, :images
  def initialize
    @images = {}
  end
end

def get_text(dom, tag)
  if dom.elements[tag]
    dom.elements[tag].text
  else
    ''
  end
end

def parse_xmls(xmls)
  bookinfos = []
  xmls.each do |xmlstring|
    doc = REXML::Document.new(xmlstring)
    image_labels = %w(SmallImage MediumImage LargeImage)
    doc.elements.each('ItemLookupResponse/Items/Item') do |item|
      bookinfo = BookInfo.new
      bookinfo.asin = item.elements['ASIN'].text
      attr = item.elements['ItemAttributes']
      bookinfo.title = get_text(attr, 'Title')
      bookinfo.binding = get_text(attr, 'Binding')
      bookinfo.author = get_text(attr, 'Author')
      bookinfo.publisher = get_text(attr, 'Publisher')
      bookinfo.publicationDate = get_text(attr, 'PublicationDate')

      image_labels.each do |image_label|
        next unless item.elements[image_label]
        image = ImageInfo.new
        imgtag = item.elements[image_label]
        image.url = imgtag.elements['URL'].text
        image.width = imgtag.elements['Width'].text.to_i
        image.height = imgtag.elements['Height'].text.to_i
        bookinfo.images[image_label] = image
      end
      bookinfos << bookinfo
    end
  end
  bookinfos
end

def get_xmls
  xmls = []
  Dir.glob('xmls/*.xml').each do |file|
    open(file, 'r') do |io|
      xmls << io.read
    end
  end
  xmls
end

xmls = get_xmls
t = Time.now
bookinfos = parse_xmls(xmls)
duration = (Time.now.to_f - t.to_f)
puts "xml数: #{xmls.size}"
puts "book数: #{bookinfos.size}"
puts "parse時間: #{duration}"
